controllers/routing/routing_reconcile_resources.go

```diff
@@ -17,8 +17,17 @@ import (
 func (r *PlatformRoutingController) reconcileResources(ctx context.Context, target *unstructured.Unstructured) error {
 	// TODO shouldn't we make it a predicate for ctrl watch instead?
 	_, exportModeFound := extractExportModes(target)
+
 	if !exportModeFound {
-		return nil
+		r.log.Info("No export mode found for target")
+
+		// deleting resources in case annotation previously existed
+		err := r.deleteOwnedResources(ctx, target, routingResourceGVKs(spi.ExternalRoute))
+		if err != nil {
```

Comment:
How about slightly different take on the problem. 

We know all rouing types, so how about constructing a slice with a difference, meaning route types that are not defined for a reconciled resource? Passing an empty slice would give us all types, for all the others we would return the difference (slice without routing types currently in use).

Having this information we could easily make a general loop/func:

```golang
exportModes, _ := extractExportModes(target)

unusedRouteTypes := spi.UnusedRouteTypes(exportModes...) 

for _, unusedRouteType := unusedRouteTypes  {
	 r.deleteOwnedResources(ctx, target, routingResourceGVKs(unusedRouteType))
}
```

similar to what you try to do in `exportService` func when looping over `spi.AllRouteTypes()`, but we could move it up and iterate over the remaining, unused routes uniformly. I also believe this logic shouldn't belong to `exportService` but top-level reconcile (we could call it `removeUnusedRoutingResource`). Then `reconcileResource` (renamed to `createRoutingResources`) would only create needed routing resources afterwards.

So essentially: 

```golang
reconcilers := []platformctrl.SubReconcileFunc{
	r.removeUnusedRoutingResource, 
	r.createRoutingResources,
}
```

We are cleaning up our resources in 3 different places now. With this approach, we can call it once in reconciliation regardless of the reason why it is invoked. In case of deletion of watched resource, we simply can swap `unusedRouteTypes` to `spi.AllRouteTypes()`  (which we can also check inside `removeUnusedRoutingResource`).

And one more thing - we have overlapping resources for both modes, `VirtualService`. In case of switching from "public;external" to one of them we will first remove both VS for no good reason. I think we should add another label when creating our resource - `routing.opendatahub.io/type` (to not conflate with annotation `routing.opendatahub.io/export-mode` I think we should use `type` instead of `export-mode` here). This way we can narrow the selectors to the exact resources we want to remove. I think we should also capture it as a test.

Then at the end of the reconciliation, if the resource is about to be deleted, and previous reconcilers didn't fail, we would only remove the finalizer. 

I hope I am making sense :) WDYT @cam-garrison @aslakknutsen ?

Reply:
Retrospecting on the `extractExportModes` code I think it should just return a slice. I  am not convinced that `bool` indicating if there's an annotation or not has any value for us.

Reply:
Speaking with @aslakknutsen we realized what we need here is a binding resource to become an owner of all the resources we create for a given path (route type) and we would delete only this one finding it by the right label... `FeatureTracker`? :) we should plan to make it reusable in the long term.

Reply:
[6cf5f20](https://github.com/maistra/odh-platform/pull/48/commits/6cf5f20415cf4395ab9c4eecf3aba41732040bbf)

I think I addressed the above review, think it is better now. Re-using `FeatureTracker`  for this would be cool :)